src/zlib.zig

```diff
@@ -901,3 +945,277 @@ pub const ZlibCompressorArrayList = struct {
+        // the ordering of this logic is a bit abstract because we dont know ahead of time how large 'bytes' will be.
+        // additionally, if the first byte is "correct" but the second is not, we don't want to lose it from it being consumed.
+        if (this.mode == .UNZIP) {
+            var redd: usize = 0;
+            this.do_inflate_loop = false;
+
+            if (bytes.len > 0) {
+                this.next_expected_header_byte = state.next_in;
+            }
+            if (this.gzip_id_bytes_read == 0) {
+                if (this.next_expected_header_byte == null) {
+                    return 0;
+                }
+
+                if (this.next_expected_header_byte.?[redd] == GZIP_HEADER_ID1) {
+                    this.gzip_id_bytes_read += 1;
+                    redd += 1;
+                    // next_expected_header_byte++;
+
+                    if (bytes.len == 1) {
+                        // The only available byte was already read.
+                        return 1;
+                    }
+                } else {
+                    // the stream did not match the gzip header, bail.
+                    this.mode = .INFLATE;
+                    return 0;
+                }
+            }
+            if (this.gzip_id_bytes_read == 1) {
+                if (this.next_expected_header_byte == null) {
+                    return 0;
+                }
+
+                if (this.next_expected_header_byte.?[redd] == GZIP_HEADER_ID2) {
+                    this.gzip_id_bytes_read += 1;
+                    redd += 1;
+                    // next_expected_header_byte++;
+
+                    // the gzip header was found. send the header to inflate() and tell writeAll how much we read to do this detection
+                    // if we continued to doWork right now GZIP_HEADER_ID2 might get processed twice.
+                    this.mode = .GUNZIP;
+                    {
+                        const header = &[_]u8{ GZIP_HEADER_ID1, GZIP_HEADER_ID2 };
+                        state.next_in = header.ptr;
+                        state.avail_in = @intCast(header.len);
+                        var out: [0]u8 = undefined;
+                        state.avail_out = 0;
+                        state.next_out = &out; // passing a null pointer here causes it to return Z_STREAM_ERROR so we send zero-length instead.
+                        const ret = inflate(state, this.flush);
+                        bun.assert(ret == .Ok);
+                    }
+                } else {
+                    // There is no actual difference between INFLATE and INFLATERAW (after initialization).
+                    // the stream only partially matched the gzip header, bail.
+                    this.mode = .INFLATE;
+                }
+                return redd;
+            }
+
+            bun.assert(false); // invalid number of gzip magic number bytes read
```

Comment:
on windows, this is going to panic at runtime instead of throw an error

should it return an error?

Reply:
its unreachable but i didnt want to use `unreachable;`